components/function-controller/pkg/controller/function/function_controller.go

```diff
@@ -373,167 +371,154 @@ func generateFunctionHash(fnCm *corev1.ConfigMap) (string, error) {
-	if err != nil && errors.IsNotFound(err) {
-		log.Info("Creating Knative Build", "namespace", deployBuild.Namespace, "name", deployBuild.Name)
-		err = r.Create(context.TODO(), deployBuild)
-		if err != nil {
-			return err
-		}
-
-		err = r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionBuilding)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
-		return nil
-
-	} else if err != nil {
-		log.Error(err, "Error while trying to create Knative Build", "namespace", deployBuild.Namespace, "name", deployBuild.Name)
-		return err
+	if err := controllerutil.SetControllerReference(fn, desiredTr, r.scheme); err != nil {
+		return nil, fmt.Errorf("error setting controller reference: %s", err)
 	}
 
-	// Update Build object
-	if !reflect.DeepEqual(deployBuild.Spec, foundBuild.Spec) && !compareBuildImages(foundBuild, imageName) {
-		err := r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionUpdating)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
+	// note: a TaskRun must be recreated upon changes, so we only support
+	// the creation of new TaskRuns, not the update after creation.
+	return r.getOrCreateFunctionBuildTaskRun(desiredTr, fn)
+}
 
-		// create new Build with the new updated image
-		log.Info("Creating new Knative Build", "namespace", deployBuild.Namespace, "name", deployBuild.Name)
-		err = r.Create(context.TODO(), deployBuild)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				return nil
-			}
-			return err
-		}
+// getOrCreateFunctionBuildTaskRun returns the existing Function build TaskRun
+// or creates it from the given desired state if it does not exist.
+func (r *ReconcileFunction) getOrCreateFunctionBuildTaskRun(desiredTr *tektonv1alpha1.TaskRun,
+	fn *serverlessv1alpha1.Function) (*tektonv1alpha1.TaskRun, error) {
 
-		log.Info("Updated Knative Build", "namespace", deployBuild.Namespace, "name", deployBuild.Name)
+	ctx := context.TODO()
 
-		return nil
-	}
+	currentTr := &tektonv1alpha1.TaskRun{}
+	err := r.Get(ctx,
+		client.ObjectKey{
+			Name:      desiredTr.Name,
+			Namespace: desiredTr.Namespace,
+		},
+		currentTr,
+	)
 
-	return nil
-}
+	switch {
+	case errors.IsNotFound(err):
+		// TODO(antoineco): a Kubernetes event would be more suitable than a log entry
+		log.Info("Creating Function build TaskRun", "namespace", desiredTr.Namespace, "name", desiredTr.Name)
 
-// compareBuildImages returns whether two builds are equal.
-func compareBuildImages(foundBuild *buildv1alpha1.Build, imageName string) bool {
-	if foundBuild.Spec.Template != nil && len(foundBuild.Spec.Template.Arguments) > 0 {
-		args := foundBuild.Spec.Template.Arguments
-		for _, arg := range args {
-			if arg.Name == "IMAGE" && arg.Value == imageName {
-				return true
-			}
+		if err := r.Create(ctx, desiredTr); err != nil {
+			return nil, err
+		}
+
+		// TODO(antoineco): it would be enough to compute the status ONCE, at the end of the sync
```

Comment:
we should change the status resource to allow multiple conditions. then we could have functions in status building and simultaneously serving. If we do so we would probably keep the current updates along the way

Reply:
Controllers (in the Kubernetes core at least) usually set status by observation at the _end_ of the reconciliation. Things are created/updated/deleted in the main logic, and then `setStatus()` (example name) observes the current state of the world to compute the entire status, ideally from informers cache for speed.

Reply:
Moving the discussion to #5673